File: sim/ros2_ws/src/ai_ship_robot_gazebo/scripts/scripted_drive.py
#!/usr/bin/env python3

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import rclpy
import yaml
from geometry_msgs.msg import Twist
from rclpy.node import Node


COMMAND_AXES = {
    "forward": ("linear_x", 1.0),
    "backward": ("linear_x", -1.0),
    "left": ("linear_y", 1.0),
    "right": ("linear_y", -1.0),
    "yaw_left": ("angular_z", 1.0),
    "yaw_right": ("angular_z", -1.0),
}


@dataclass(frozen=True)
class ScenarioStep:
    duration_sec: float
    linear_x: float
    linear_y: float
    angular_z: float


class ScriptedDrive(Node):
    def __init__(self) -> None:
        super().__init__("scripted_drive")
        self.declare_parameter("cmd_vel_topic", "cmd_vel")
        self.declare_parameter("scenario_file", "")
        self.declare_parameter("start_delay_sec", 0.0)
        self.declare_parameter("loop", False)
        self.declare_parameter("publish_rate", 10.0)

        self.cmd_vel_topic = str(self.get_parameter("cmd_vel_topic").value)
        self.scenario_file = Path(str(self.get_parameter("scenario_file").value))
        self.start_delay_sec = float(self.get_parameter("start_delay_sec").value)
        self.loop_enabled = bool(self.get_parameter("loop").value)
        self.publish_rate = float(self.get_parameter("publish_rate").value)
        self.publisher = self.create_publisher(Twist, self.cmd_vel_topic, 10)
        self.steps = self._load_scenario(self.scenario_file)

    def _load_scenario(self, scenario_file: Path) -> list[ScenarioStep]:
        # 設定ファイルの構造を起動直後に厳密検証し、実行中エラーを避ける。
        if not scenario_file.is_file():
            raise ValueError(f"Scenario file not found: {scenario_file}")

        with scenario_file.open("r", encoding="utf-8") as stream:
            data = yaml.safe_load(stream)

        if not isinstance(data, dict) or not isinstance(data.get("steps"), list):
            raise ValueError("Scenario YAML must contain a top-level 'steps' list.")

        steps: list[ScenarioStep] = []
        for index, raw_step in enumerate(data["steps"], start=1):
            steps.append(self._parse_step(index, raw_step))

        if not steps:
            raise ValueError("Scenario YAML must define at least one step.")
        return steps

    def _parse_step(self, index: int, raw_step: Any) -> ScenarioStep:
        # 各ステップは速度の持ち越しを防ぐため、毎回ゼロ初期化から合成する。
        if not isinstance(raw_step, dict):
            raise ValueError(f"steps[{index}] must be a mapping.")

        try:
            duration_sec = float(raw_step["duration_sec"])
        except KeyError as exc:
            raise ValueError(f"steps[{index}] is missing duration_sec.") from exc
        except (TypeError, ValueError) as exc:
            raise ValueError(f"steps[{index}].duration_sec must be numeric.") from exc

        if duration_sec <= 0.0:
            raise ValueError(f"steps[{index}].duration_sec must be > 0.")

        commands = raw_step.get("commands")
        if not isinstance(commands, list) or not commands:
            raise ValueError(f"steps[{index}].commands must be a non-empty list.")

        components = {"linear_x": 0.0, "linear_y": 0.0, "angular_z": 0.0}
        used_axes: dict[str, str] = {}
        stop_used = False

        for command_index, raw_command in enumerate(commands, start=1):
            self._apply_command(index, command_index, raw_command, components, used_axes)
            stop_used = stop_used or raw_command.get("type") == "stop"

        if stop_used and len(commands) != 1:
            raise ValueError(f"steps[{index}] stop must be the only command in the step.")

        return ScenarioStep(duration_sec=duration_sec, **components)

    def _apply_command(
        self,
        step_index: int,
        command_index: int,
        raw_command: Any,
        components: dict[str, float],
        used_axes: dict[str, str],
    ) -> None:
        # 同一軸の正負コマンド競合をここで止め、曖昧なTwist生成を禁止する。
        if not isinstance(raw_command, dict):
            raise ValueError(f"steps[{step_index}].commands[{command_index}] must be a mapping.")

        command_type = raw_command.get("type")
        if command_type == "stop":
            return
        if command_type not in COMMAND_AXES:
            raise ValueError(f"Unsupported command type at steps[{step_index}]: {command_type}")

        try:
            speed = float(raw_command["speed"])
        except KeyError as exc:
            raise ValueError(f"steps[{step_index}].commands[{command_index}] is missing speed.") from exc
        except (TypeError, ValueError) as exc:
            raise ValueError(f"steps[{step_index}].commands[{command_index}].speed must be numeric.") from exc

        if speed < 0.0:
            raise ValueError(f"steps[{step_index}].commands[{command_index}].speed must be >= 0.")

        axis_name, direction = COMMAND_AXES[command_type]
        if axis_name in used_axes:
            raise ValueError(
                f"steps[{step_index}] cannot combine {used_axes[axis_name]} and {command_type} on {axis_name}."
            )
        used_axes[axis_name] = command_type
        components[axis_name] = direction * speed

    def publish_twist(self, linear_x: float, linear_y: float, angular_z: float) -> None:
        twist = Twist()
        twist.linear.x = linear_x
        twist.linear.y = linear_y
        twist.angular.z = angular_z
        self.publisher.publish(twist)

    def stop(self) -> None:
        self.publish_twist(0.0, 0.0, 0.0)

    def wait_for_sim_duration(self, duration_sec: float) -> bool:
        # 自動運転の待機はsim time基準で行い、Gazeboの進行に同期させる。
        start_time = None
        target_duration_ns = int(duration_sec * 1_000_000_000)

        while rclpy.ok():
            rclpy.spin_once(self, timeout_sec=0.01)
            current_time = self.get_clock().now()
            if current_time.nanoseconds == 0:
                continue
            if start_time is None:
                # /clock受信前の0時刻を除外し、起動時に観測した有効なsim timeを待機起点にする。
                start_time = current_time
                continue
            if (current_time - start_time).nanoseconds >= target_duration_ns:
                return True

        return False

    def format_step_command(self, step: ScenarioStep) -> str:
        # 実行中ログで確認しやすいよう、各軸の指令値を固定形式で整形する。
        return (
            f"linear_x={step.linear_x:.3f}, linear_y={step.linear_y:.3f}, "
            f"angular_z={step.angular_z:.3f}, duration={step.duration_sec:.2f}s"
        )

    def run(self) -> None:
        # publish_rateは送信周期として使い、ステップ継続時間そのものはsim timeで判定する。
        if self.publish_rate <= 0.0:
            raise ValueError("publish_rate must be > 0.")
        if self.start_delay_sec > 0.0:
            self.get_logger().info(
                f"Waiting {self.start_delay_sec:.2f} sec before scenario start (sim time)"
            )
            if not self.wait_for_sim_duration(self.start_delay_sec):
                return

        period_sec = 1.0 / self.publish_rate
        keep_running = True
        while rclpy.ok() and keep_running:
            # ループ実行時も各周回の先頭を記録し、どのシナリオ周回か追跡できるようにする。
            self.get_logger().info("Starting scripted drive scenario")
            for step_index, step in enumerate(self.steps, start=1):
                deadline = self.get_clock().now().nanoseconds + int(step.duration_sec * 1_000_000_000)
                self.get_logger().info(f"Step {step_index}: {self.format_step_command(step)}")
                while rclpy.ok():
                    if not rclpy.ok():
                        return
                    self.publish_twist(step.linear_x, step.linear_y, step.angular_z)
                    rclpy.spin_once(self, timeout_sec=period_sec)
                    current_time = self.get_clock().now().nanoseconds
                    if current_time != 0 and current_time >= deadline:
                        break
                self.stop()
                self.get_logger().info(f"Step {step_index}: stop")
            keep_running = self.loop_enabled


def main() -> int:
    rclpy.init()
    node: ScriptedDrive | None = None

    try:
        # 例外時も停止publishを試みるため、Node生成後はfinallyで必ずstopを呼ぶ。
        node = ScriptedDrive()
        node.run()
        return 0
    except KeyboardInterrupt:
        return 0
    except Exception as exc:
        print(f"scripted_drive failed: {exc}", flush=True)
        return 1
    finally:
        if node is not None:
            try:
                node.stop()
            except Exception:
                pass
            node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    raise SystemExit(main())

File: sim/ros2_ws/src/ai_ship_robot_gazebo/src/scripted_drive.cpp
#include "ai_ship_robot_gazebo/scripted_drive_core.hpp"

#include <rclcpp/rclcpp.hpp>

#include <geometry_msgs/msg/twist.hpp>
#include <nav_msgs/msg/odometry.hpp>

#include <algorithm>
#include <chrono>
#include <cmath>
#include <exception>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>

using namespace std::chrono_literals;

namespace ai_ship_robot_gazebo
{
namespace
{

constexpr double kPi = 3.14159265358979323846;
constexpr double kMinControlRemainingSec = 0.05;
constexpr double kDefaultPositionTolerance = 0.05;
constexpr double kDefaultYawTolerance = kPi / 180.0;

struct PlanarPose
{
  double x{};
  double y{};
  double yaw{};
};

struct AbsoluteMoveToPoseTarget
{
  double x{};
  double y{};
  double position_tolerance{};
  double yaw{};
  double yaw_tolerance{};
};

double normalize_angle(const double angle)
{
  auto normalized = std::fmod(angle + kPi, 2.0 * kPi);
  if (normalized < 0.0) {
    normalized += 2.0 * kPi;
  }
  return normalized - kPi;
}

double yaw_from_odometry(const nav_msgs::msg::Odometry & odometry)
{
  const auto & orientation = odometry.pose.pose.orientation;
  const auto siny_cosp = 2.0 * (orientation.w * orientation.z + orientation.x * orientation.y);
  const auto cosy_cosp = 1.0 - 2.0 * (orientation.y * orientation.y + orientation.z * orientation.z);
  return std::atan2(siny_cosp, cosy_cosp);
}

}  // namespace

class ScriptedDrive : public rclcpp::Node
{
public:
  ScriptedDrive()
  : Node("scripted_drive")
  {
    declare_parameter<std::string>("cmd_vel_topic", "cmd_vel");
    declare_parameter<std::string>("odom_topic", "odom");
    declare_parameter<std::string>("scenario_file", "");
    declare_parameter<double>("start_delay_sec", 0.0);
    declare_parameter<double>("publish_rate", 10.0);
    declare_parameter<double>("max_linear_speed", 0.5);
    declare_parameter<double>("max_angular_speed", 1.0);

    cmd_vel_topic_ = get_parameter("cmd_vel_topic").as_string();
    odom_topic_ = get_parameter("odom_topic").as_string();
    scenario_file_ = get_parameter("scenario_file").as_string();
    start_delay_sec_ = get_parameter("start_delay_sec").as_double();
    publish_rate_ = get_parameter("publish_rate").as_double();
    max_linear_speed_ = get_parameter("max_linear_speed").as_double();
    max_angular_speed_ = get_parameter("max_angular_speed").as_double();

    publisher_ = create_publisher<geometry_msgs::msg::Twist>(cmd_vel_topic_, rclcpp::QoS(10));
    odom_subscription_ = create_subscription<nav_msgs::msg::Odometry>(
      odom_topic_, rclcpp::QoS(10), [this](const nav_msgs::msg::Odometry::SharedPtr message) {
        latest_pose_ = PlanarPose{message->pose.pose.position.x, message->pose.pose.position.y, yaw_from_odometry(*message)};
      });
    scenario_ = load_scenario_file(scenario_file_);
  }

  void run()
  {
    if (publish_rate_ <= 0.0) {
      throw std::invalid_argument("publish_rate must be > 0.");
    }
    if (max_linear_speed_ <= 0.0) {
      throw std::invalid_argument("max_linear_speed must be > 0.");
    }
    if (max_angular_speed_ <= 0.0) {
      throw std::invalid_argument("max_angular_speed must be > 0.");
    }
    if (!wait_for_active_sim_time()) {
      return;
    }
    if (start_delay_sec_ > 0.0) {
      RCLCPP_INFO(get_logger(), "Waiting %.2f sec before scenario start (sim time)", start_delay_sec_);
      if (!wait_for_sim_duration(start_delay_sec_)) {
        return;
      }
    }

    const auto period = std::chrono::duration<double>(1.0 / publish_rate_);
    RCLCPP_INFO(get_logger(), "Starting scripted drive scenario");
    run_steps(period);
  }

  void stop()
  {
    publish_twist(0.0, 0.0, 0.0);
  }

private:
  std::optional<rclcpp::Time> wait_for_active_sim_time()
  {
    rclcpp::WallRate rate(100.0);

    // use_sim_timeの/clockを受信するまで待ち、0秒基準でdurationを計算して即終了する誤動作を防ぐ。
    while (rclcpp::ok()) {
      rclcpp::spin_some(shared_from_this());
      const auto current_time = get_clock()->now();
      if (current_time.nanoseconds() != 0) {
        return current_time;
      }
      rate.sleep();
    }
    return std::nullopt;
  }

  bool wait_for_sim_duration(const double duration_sec)
  {
    const auto active_time = wait_for_active_sim_time();
    if (!active_time) {
      return false;
    }
    auto start_time = *active_time;
    const auto target_duration = rclcpp::Duration::from_seconds(duration_sec);
    rclcpp::WallRate rate(100.0);

    // Gazeboの時刻進行に同期して待機し、real_time_factorの影響を正しく反映する。
    while (rclcpp::ok()) {
      rclcpp::spin_some(shared_from_this());
      const auto current_time = get_clock()->now();
      if (current_time.nanoseconds() == 0) {
        rate.sleep();
        continue;
      }
      if (current_time < start_time) {
        start_time = current_time;
        rate.sleep();
        continue;
      }
      if ((current_time - start_time) >= target_duration) {
        return true;
      }
      rate.sleep();
    }
    return false;
  }

  void run_steps(const std::chrono::duration<double> period)
  {
    for (std::size_t index = 0; rclcpp::ok() && index < scenario_.steps.size(); ++index) {
      const auto & step = scenario_.steps[index];
      RCLCPP_INFO(get_logger(), "Step %zu: %s", index + 1, format_step_command(step).c_str());

      bool completed = false;
      if (step.type == ScenarioStepType::Velocity) {
        completed = run_velocity_step(step, period);
      } else if (step.type == ScenarioStepType::MoveToPose) {
        completed = run_move_to_pose_step(step, period);
      }

      if (!completed) {
        stop();
        throw std::runtime_error("Step " + std::to_string(index + 1) + " failed.");
      }
      stop();
      RCLCPP_INFO(get_logger(), "Step %zu: stop", index + 1);
    }
  }

  bool prepare_step_time(const ScenarioStep & step, rclcpp::Time & step_start_time, rclcpp::Time & deadline)
  {
    const auto active_step_time = wait_for_active_sim_time();
    if (!active_step_time) {
      return false;
    }
    step_start_time = *active_step_time;
    deadline = step_start_time + rclcpp::Duration::from_seconds(step.duration_sec);
    return true;
  }

  void adjust_time_if_reset(
    const ScenarioStep & step, const rclcpp::Time & current_time, rclcpp::Time & step_start_time,
    rclcpp::Time & deadline)
  {
    if (current_time < step_start_time) {
      step_start_time = current_time;
      deadline = current_time + rclcpp::Duration::from_seconds(step.duration_sec);
    }
  }

  bool run_velocity_step(const ScenarioStep & step, const std::chrono::duration<double> period)
  {
    rclcpp::Time step_start_time;
    rclcpp::Time deadline;
    if (!prepare_step_time(step, step_start_time, deadline)) {
      return false;
    }

    // 速度stepは従来通り一定Twistを出し、終了判定だけGazebo時刻で行う。
    rclcpp::WallRate rate(1.0 / period.count());
    while (rclcpp::ok()) {
      publish_twist(step.linear_x, step.linear_y, step.angular_z);
      rclcpp::spin_some(shared_from_this());
      const auto current_time = get_clock()->now();
      if (current_time.nanoseconds() == 0) {
        rate.sleep();
        continue;
      }
      adjust_time_if_reset(step, current_time, step_start_time, deadline);
      if (current_time >= deadline) {
        return true;
      }
      rate.sleep();
    }
    return false;
  }

  bool run_move_to_pose_step(const ScenarioStep & step, const std::chrono::duration<double> period)
  {
    rclcpp::Time step_start_time;
    rclcpp::Time deadline;
    if (!prepare_step_time(step, step_start_time, deadline)) {
      return false;
    }
    const auto target = resolve_move_to_pose_target(step);
    if (!target) {
      return false;
    }

    // duration_secを速度算出用の目標到達時間として使い、到達判定は許容誤差だけで行う。
    rclcpp::WallRate rate(1.0 / period.count());
    while (rclcpp::ok()) {
      rclcpp::spin_some(shared_from_this());
      const auto current_time = get_clock()->now();
      if (current_time.nanoseconds() == 0 || !latest_pose_) {
        rate.sleep();
        continue;
      }
      adjust_time_if_reset(step, current_time, step_start_time, deadline);
      const auto dx = target->x - latest_pose_->x;
      const auto dy = target->y - latest_pose_->y;
      const auto distance = std::hypot(dx, dy);
      const auto yaw_error = normalize_angle(target->yaw - latest_pose_->yaw);
      if (distance <= target->position_tolerance && std::abs(yaw_error) <= target->yaw_tolerance) {
        return true;
      }
      const auto remaining_sec = std::max((deadline - current_time).seconds(), kMinControlRemainingSec);
      const auto cos_yaw = std::cos(latest_pose_->yaw);
      const auto sin_yaw = std::sin(latest_pose_->yaw);
      auto linear_x = (cos_yaw * dx + sin_yaw * dy) / remaining_sec;
      auto linear_y = (-sin_yaw * dx + cos_yaw * dy) / remaining_sec;
      const auto angular_z = std::clamp(yaw_error / remaining_sec, -max_angular_speed_, max_angular_speed_);
      clamp_linear_velocity(linear_x, linear_y);
      publish_twist(linear_x, linear_y, angular_z);
      rate.sleep();
    }
    return false;
  }

  std::optional<AbsoluteMoveToPoseTarget> resolve_move_to_pose_target(const ScenarioStep & step)
  {
    if (!wait_for_odometry()) {
      return std::nullopt;
    }

    const auto start_pose = *latest_pose_;
    AbsoluteMoveToPoseTarget target;
    target.x = start_pose.x;
    target.y = start_pose.y;
    target.yaw = start_pose.yaw;
    target.position_tolerance = step.move_to_pose.has_position ? step.move_to_pose.position.tolerance : kDefaultPositionTolerance;
    target.yaw_tolerance = step.move_to_pose.has_yaw ? step.move_to_pose.yaw.tolerance : kDefaultYawTolerance;
    if (step.move_to_pose.mode == MoveToPoseMode::Absolute) {
      if (step.move_to_pose.has_position) {
        target.x = step.move_to_pose.position.x;
        target.y = step.move_to_pose.position.y;
      }
      if (step.move_to_pose.has_yaw) {
        target.yaw = step.move_to_pose.yaw.yaw;
      }
      return target;
    }

    // relはstep開始時のbase姿勢を基準にし、相対座標をodom座標の絶対目標へ変換する。
    const auto cos_yaw = std::cos(start_pose.yaw);
    const auto sin_yaw = std::sin(start_pose.yaw);
    if (step.move_to_pose.has_position) {
      target.x = start_pose.x + cos_yaw * step.move_to_pose.position.x - sin_yaw * step.move_to_pose.position.y;
      target.y = start_pose.y + sin_yaw * step.move_to_pose.position.x + cos_yaw * step.move_to_pose.position.y;
    }
    if (step.move_to_pose.has_yaw) {
      target.yaw = normalize_angle(start_pose.yaw + step.move_to_pose.yaw.yaw);
    }
    return target;
  }

  bool wait_for_odometry()
  {
    rclcpp::WallRate rate(100.0);
    while (rclcpp::ok() && !latest_pose_) {
      rclcpp::spin_some(shared_from_this());
      rate.sleep();
    }
    return latest_pose_.has_value();
  }

  void clamp_linear_velocity(double & linear_x, double & linear_y) const
  {
    const auto speed = std::hypot(linear_x, linear_y);
    if (speed <= max_linear_speed_) {
      return;
    }

    const auto scale = max_linear_speed_ / speed;
    linear_x *= scale;
    linear_y *= scale;
  }

  void publish_twist(const double linear_x, const double linear_y, const double angular_z)
  {
    geometry_msgs::msg::Twist twist;
    twist.linear.x = linear_x;
    twist.linear.y = linear_y;
    twist.angular.z = angular_z;
    publisher_->publish(twist);
  }

  std::string cmd_vel_topic_;
  std::string odom_topic_;
  std::string scenario_file_;
  double start_delay_sec_{};
  double publish_rate_{};
  double max_linear_speed_{};
  double max_angular_speed_{};
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscription_;
  std::optional<PlanarPose> latest_pose_;
  ScenarioDefinition scenario_;
};

}  // namespace ai_ship_robot_gazebo

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  std::shared_ptr<ai_ship_robot_gazebo::ScriptedDrive> node;

  try {
    node = std::make_shared<ai_ship_robot_gazebo::ScriptedDrive>();
    node->run();
  } catch (const std::exception & exc) {
    if (node) {
      RCLCPP_ERROR(node->get_logger(), "scripted_drive failed: %s", exc.what());
    } else {
      fprintf(stderr, "scripted_drive failed: %s\n", exc.what());
    }
    if (node) {
      node->stop();
    }
    rclcpp::shutdown();
    return 1;
  }

  if (node) {
    node->stop();
  }
  rclcpp::shutdown();
  return 0;
}
